### ISSUE TITLE
Adreno(TM) 506 opencl cost too much time

### ISSUE THREAD

**Issue by mathpopo** [ARM-software/ComputeLibrary#846]
@sanpar02 Adreno(TM) 506 opencl cost too much time ,because which need Adreno self optimization?

/benchmark_graph_mnist --example_args=--target=CL <
Version = arm_compute_version=v20.05 Build options: {'arch': 'arm64-v8a', 'neon': '1', 'opencl': '1', 'embed_kernels': '1', 'extra_cxx_flags': '-fPIC', 'benchmark_tests': '0', 'validation_tests': '0', 'os': 'android'} Git hash=b'6a7771e460abeac7d401d6d38a0fcf0a0d2c3cbe'
CommandLine = ./benchmark_graph_mnist --example_args=--target=CL
CL_DEVICE_VERSION = OpenCL 2.0 Adreno(TM) 506
Iterations = 1
Running [0] 'Examples/benchmark_graph_mnist'
Threads : 1
Target : CL
Data type : F32
Data layout : NHWC
Tuner enabled? : false
Cache enabled? : true
Tuner mode : Normal
Tuner file :
Fast math enabled? : false

Wall clock/Wall clock time: AVG=4061.0000 us
Executed 1 test(s) (1 passed, 0 expected failures, 0 failed, 0 crashed, 0 disabled) in 7 second(s)
HWBKK-Q:/data/local/tmp $ ./benchmark_graph_mnist
Version = arm_compute_version=v20.05 Build options: {'arch': 'arm64-v8a', 'neon': '1', 'opencl': '1', 'embed_kernels': '1', 'extra_cxx_flags': '-fPIC', 'benchmark_tests': '0', 'validation_tests': '0', 'os': 'android'} Git hash=b'6a7771e460abeac7d401d6d38a0fcf0a0d2c3cbe'
CommandLine = ./benchmark_graph_mnist
CL_DEVICE_VERSION = OpenCL 2.0 Adreno(TM) 506
Iterations = 1
Running [0] 'Examples/benchmark_graph_mnist'
Threads : 1
Target : NEON
Data type : F32
Data layout : NHWC
Tuner enabled? : false
Cache enabled? : false
Tuner mode : Normal
Tuner file :
Fast math enabled? : false

Wall clock/Wall clock time: AVG=1895.0000 us
Executed 1 test(s) (1 passed, 0 expected failures, 0 failed, 0 crashed, 0 disabled) in 0 second(s)

**Comment by sanpar02** [comment]
Hi @mathpopo 

If the benchmark takes too much time on non-Arm GPUs, the GPU itself could be certainly one of the reasons since Compute Library is optimized for Arm CPUs and GPUs.

I'll try to assign this ticket to another colleague who has more insights on performance.

Thanks.

**Comment by sanpar02** [comment]
Hi @mathpopo 

It might be worth trying to give a command-line option `--iteration=N` to repeat the benchmark N times and see how the execution time changes, but I wouldn't expect too much difference.

Also, I have double-checked with another colleague and the answer was pretty similar. So, it is difficult for us to comment on or try to improve the performance using non-Arm GPUs.

I hope it helps.

Thanks.

**Comment by mathpopo** [comment]
@sanpar02 very  thanks,you mean Arm GPUs--->
                {"Mali-T760", 6.83f},
                {"Mali-T860", 6.83f},
                {"Mali-T880", 6.83f},
                {"Mali-G51", 6.83f},
                {"Mali-G52", 6.83f},
                {"Mali-G71", 31.61f},
                {"Mali-G72", 31.61f},
                {"Mali-G76", 31.61f},--------------------------------------------------------------above?
                {"Adreno (TM) 505", 3.19f},
                {"Adreno (TM) 506", 4.74f},
                {"Adreno (TM) 512", 14.23f},
                {"Adreno (TM) 530", 25.40f},
                {"Adreno (TM) 540", 42.74f},
                {"Adreno (TM) 615", 16.77f},
                {"Adreno (TM) 616", 18.77f},
                {"Adreno (TM) 618", 18.77f},
                {"Adreno (TM) 630", 42.74f},
                {"Adreno (TM) 640", 42.74f},

**Comment by mathpopo** [comment]
Mali-T760  cannot run

**Comment by mathpopo** [comment]
https://github.com/ARM-software/ComputeLibrary/issues/842  show G71 opencl run error

**Comment by mathpopo** [comment]
HWMOA-M:/data/local/tmp $ ./benchmark_graph_mnist --example_args=--target=CL
[CORE][15-07-2020 05:42:10][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:10][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:10][INFO] Can't find valid Mali GPU. Target is set to default.
Version = arm_compute_version=v20.05 Build options: {'Werror': '1', 'debug': '1', 'neon': '1', 'opencl': '1', 'gles_compute': '1', 'examples': '1', 'os': 'android', 'mali': '1', 'arch': 'armv7a'} Git hash=b'6a7771e460abeac7d401d6d38a0fcf0a0d2c3cbe'
CommandLine = ./benchmark_graph_mnist --example_args=--target=CL
CL_DEVICE_VERSION = OpenCL 1.2
Iterations = 1
[CORE][15-07-2020 05:42:10][INFO] Can't find valid Mali GPU. Target is set to default.
Running [0] 'Examples/benchmark_graph_mnist'
Threads : 1
Target : CL
Data type : F32
Data layout : NHWC
Tuner enabled? : false
Cache enabled? : true
Tuner mode : Normal
Tuner file :
Fast math enabled? : false

[GRAPH][15-07-2020 05:42:10][INFO] Running mutating pass : NodeFusionMutator

[GRAPH][15-07-2020 05:42:10][INFO] Running mutating pass : GroupedConvolutionMutator

[GRAPH][15-07-2020 05:42:10][INFO] Running mutating pass : InPlaceOperationMutator

[GRAPH][15-07-2020 05:42:10][INFO] Running mutating pass : DepthConcatSubTensorMutator

[GRAPH][15-07-2020 05:42:10][INFO] Running mutating pass : SplitLayerSubTensorMutator

[GRAPH][15-07-2020 05:42:10][INFO] Running mutating pass : NodeExecutionMethodMutator

[CORE][15-07-2020 05:42:11][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:13][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:13][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:13][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:13][INFO] Can't find valid Mali GPU. Target is set to default.
[GRAPH][15-07-2020 05:42:15][INFO] Instantiated Conv0 Type: GenericConvolutionLayer Target: CL Data Type: F32 Groups: 1 Input shape: 1x28x28 Weights shape: 1x3x3x32 Output shape: 32x28x28

[CORE][15-07-2020 05:42:18][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:21][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:21][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:22][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:22][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:24][INFO] Can't find valid Mali GPU. Target is set to default.
[GRAPH][15-07-2020 05:42:24][INFO] Instantiated conv1 Type: GenericConvolutionLayer Target: CL Data Type: F32 Groups: 1 Input shape: 32x28x28 Weights shape: 32x3x3x32 Output shape: 32x28x28

[CORE][15-07-2020 05:42:24][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:25][INFO] Can't find valid Mali GPU. Target is set to default.
[GRAPH][15-07-2020 05:42:25][INFO] Instantiated maxpool1 Type: PoolingLayer Target: CL Data Type: F32 Input shape: 32x28x28 Output shape: 32x14x14 Pooling info: MAX

[CORE][15-07-2020 05:42:28][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:28][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:28][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:28][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:28][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:29][INFO] Can't find valid Mali GPU. Target is set to default.
[GRAPH][15-07-2020 05:42:29][INFO] Instantiated conv2 Type: GenericConvolutionLayer Target: CL Data Type: F32 Groups: 1 Input shape: 32x14x14 Weights shape: 32x3x3x32 Output shape: 32x14x14

[CORE][15-07-2020 05:42:30][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:30][INFO] Can't find valid Mali GPU. Target is set to default.
[GRAPH][15-07-2020 05:42:30][INFO] Instantiated maxpool2 Type: PoolingLayer Target: CL Data Type: F32 Input shape: 32x14x14 Output shape: 32x7x7 Pooling info: MAX

[CORE][15-07-2020 05:42:30][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:30][INFO] Can't find valid Mali GPU. Target is set to default.
[GRAPH][15-07-2020 05:42:31][INFO] Instantiated fc Type: FullyConnectedLayer Target: CL Data Type: F32 Input shape: 32x7x7 Weights shape: 1568x10 Output shape: 10

[CORE][15-07-2020 05:42:32][INFO] Can't find valid Mali GPU. Target is set to default.
[GRAPH][15-07-2020 05:42:32][INFO] Instantiated prob Type: SoftmaxLayer Target: CL Data Type: F32 Input shape: 10 Output shape: 10

Segmentation fault

**Comment by sanpar02** [comment]
Hi @mathpopo 

Could you rephrase your question? Did you mean that you failed to execute using T760 and G71?

The log you provided seems to say it couldn't find Mali GPU, so I'm a bit confused.

```
[CORE][15-07-2020 05:42:10][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:10][INFO] Can't find valid Mali GPU. Target is set to default.
[CORE][15-07-2020 05:42:10][INFO] Can't find valid Mali GPU. Target is set to default.
```

**Comment by mathpopo** [comment]
@sanpar02   yes ,maybe i just use mtk gpu,so many error, i look for a real G71
![123](https://user-images.githubusercontent.com/21274466/88239308-9f7b6600-ccb6-11ea-9476-23df21202ca6.jpeg)


HWMHA:/data/local/tmp $ ./benchmark_graph_mnist
Version = arm_compute_version=v20.05 Build options: {'arch': 'arm64-v8a', 'neon': '1', 'opencl': '1', 'embed_kernels': '1', 'extra_cxx_flags': '-fPIC', 'benchmark_tests': '0', 'validation_tests': '0', 'os': 'android'} Git hash=b'6a7771e460abeac7d401d6d38a0fcf0a0d2c3cbe'
CommandLine = ./benchmark_graph_mnist 
CL_DEVICE_VERSION = OpenCL 2.0 v1.r14p0-00cet0.9500166da1423e9591394b068a52b758
Iterations = 1
Running [0] 'Examples/benchmark_graph_mnist'
Threads : 1
Target : NEON
Data type : F32
Data layout : NHWC
Tuner enabled? : false
Cache enabled? : false
Tuner mode : Normal
Tuner file : 
Fast math enabled? : false

  Wall clock/Wall clock time:    AVG=5091.0000 us
Executed 1 test(s) (1 passed, 0 expected failures, 0 failed, 0 crashed, 0 disabled) in 0 second(s)
HWMHA:/data/local/tmp $ ./benchmark_graph_mnist --example_args=--target=CL     
Version = arm_compute_version=v20.05 Build options: {'arch': 'arm64-v8a', 'neon': '1', 'opencl': '1', 'embed_kernels': '1', 'extra_cxx_flags': '-fPIC', 'benchmark_tests': '0', 'validation_tests': '0', 'os': 'android'} Git hash=b'6a7771e460abeac7d401d6d38a0fcf0a0d2c3cbe'
CommandLine = ./benchmark_graph_mnist --example_args=--target=CL 
CL_DEVICE_VERSION = OpenCL 2.0 v1.r14p0-00cet0.9500166da1423e9591394b068a52b758
Iterations = 1
Running [0] 'Examples/benchmark_graph_mnist'
Threads : 1
Target : CL
Data type : F32
Data layout : NHWC
Tuner enabled? : false
Cache enabled? : true
Tuner mode : Normal
Tuner file : 
Fast math enabled? : false

  Wall clock/Wall clock time:    AVG=3138.0000 us
Executed 1 test(s) (1 passed, 0 expected failures, 0 failed, 0 crashed, 0 disabled) in 2 second(s)
HWMHA:/data/local/t

**Comment by mathpopo** [comment]
_mnist --example_args=--target=NEON --iteration=1000                         <
Version = arm_compute_version=v20.05 Build options: {'arch': 'arm64-v8a', 'neon': '1', 'opencl': '1', 'embed_kernels': '1', 'extra_cxx_flags': '-fPIC', 'benchmark_tests': '0', 'validation_tests': '0', 'os': 'android'} Git hash=b'6a7771e460abeac7d401d6d38a0fcf0a0d2c3cbe'
CommandLine = ./benchmark_graph_mnist --example_args=--target=NEON --iteration=1000 
CL_DEVICE_VERSION = OpenCL 2.0 v1.r14p0-00cet0.9500166da1423e9591394b068a52b758
Iterations = 1
Running [0] 'Examples/benchmark_graph_mnist'
Threads : 1
Target : NEON
Data type : F32
Data layout : NHWC
Tuner enabled? : false
Cache enabled? : false
Tuner mode : Normal
Tuner file : 
Fast math enabled? : false

  Wall clock/Wall clock time:    AVG=2084.0000 us
Executed 1 test(s) (1 passed, 0 expected failures, 0 failed, 0 crashed, 0 disabled) in 0 second(s)
aph_mnist --example_args=--target=CL --iteration=1000                         <
Version = arm_compute_version=v20.05 Build options: {'arch': 'arm64-v8a', 'neon': '1', 'opencl': '1', 'embed_kernels': '1', 'extra_cxx_flags': '-fPIC', 'benchmark_tests': '0', 'validation_tests': '0', 'os': 'android'} Git hash=b'6a7771e460abeac7d401d6d38a0fcf0a0d2c3cbe'
CommandLine = ./benchmark_graph_mnist --example_args=--target=CL --iteration=1000 
CL_DEVICE_VERSION = OpenCL 2.0 v1.r14p0-00cet0.9500166da1423e9591394b068a52b758
Iterations = 1
Running [0] 'Examples/benchmark_graph_mnist'
Threads : 1
Target : CL
Data type : F32
Data layout : NHWC
Tuner enabled? : false
Cache enabled? : true
Tuner mode : Normal
Tuner file : 
Fast math enabled? : false

  Wall clock/Wall clock time:    AVG=2773.0000 us
Executed 1 test(s) (1 passed, 0 expected failures, 0 failed, 0 crashed, 0 disabled) in 2 second(s)

**Comment by mathpopo** [comment]
but MALI-T760 (uname -a:   armv7l Android),i use arch=arnv7a computerlibrary ,run ,show nothing

**Comment by mathpopo** [comment]
ARM Mali:

/system/vendor/lib/egl/libGLES_mali.so
or /system/lib/egl/libGLES_mali.so

**Comment by sanpar02** [comment]
Hi @mathpopo 

What is the CPU on the device with G71? I suppose it is Cortex-A73 based on the screenshot? I haven't analyzed it thoroughly but at first glance, there is a good chance for CPU to outperform in this case because the network is a relatively small one. May I ask if there are any specific reasons/situations that make you think GPU should outperform?

In any case, if you have concerns about the absolute performance (not relative performance between two different hardware), we are more than willing to investigate.

In the case of Mali-T760, if it is not executing correctly, could I ask you to open a separate issue with more details?

Thanks.

**Comment by mathpopo** [comment]
@sanpar02   ok  ,i will open new issue

**Comment by Saikatsaha1996** [comment]
> @sanpar02 Adreno(TM) 506 opencl cost too much time ,because which need Adreno self optimization?
> 
> /benchmark_graph_mnist --example_args=--target=CL < Version = arm_compute_version=v20.05 Build options: {'arch': 'arm64-v8a', 'neon': '1', 'opencl': '1', 'embed_kernels': '1', 'extra_cxx_flags': '-fPIC', 'benchmark_tests': '0', 'validation_tests': '0', 'os': 'android'} Git hash=b'6a7771e460abeac7d401d6d38a0fcf0a0d2c3cbe' CommandLine = ./benchmark_graph_mnist --example_args=--target=CL CL_DEVICE_VERSION = OpenCL 2.0 Adreno(TM) 506 Iterations = 1 Running [0] 'Examples/benchmark_graph_mnist' Threads : 1 Target : CL Data type : F32 Data layout : NHWC Tuner enabled? : false Cache enabled? : true Tuner mode : Normal Tuner file : Fast math enabled? : false
> 
> Wall clock/Wall clock time: AVG=4061.0000 us Executed 1 test(s) (1 passed, 0 expected failures, 0 failed, 0 crashed, 0 disabled) in 7 second(s) HWBKK-Q:/data/local/tmp $ ./benchmark_graph_mnist Version = arm_compute_version=v20.05 Build options: {'arch': 'arm64-v8a', 'neon': '1', 'opencl': '1', 'embed_kernels': '1', 'extra_cxx_flags': '-fPIC', 'benchmark_tests': '0', 'validation_tests': '0', 'os': 'android'} Git hash=b'6a7771e460abeac7d401d6d38a0fcf0a0d2c3cbe' CommandLine = ./benchmark_graph_mnist CL_DEVICE_VERSION = OpenCL 2.0 Adreno(TM) 506 Iterations = 1 Running [0] 'Examples/benchmark_graph_mnist' Threads : 1 Target : NEON Data type : F32 Data layout : NHWC Tuner enabled? : false Cache enabled? : false Tuner mode : Normal Tuner file : Fast math enabled? : false
> 
> Wall clock/Wall clock time: AVG=1895.0000 us Executed 1 test(s) (1 passed, 0 expected failures, 0 failed, 0 crashed, 0 disabled) in 0 second(s)

I already build arm compute library but how can I install in my device with adreno GPU ?

**Comment by morgolock** [comment]
Hi @Saikatsaha1996 

Quote from ACL documentation

> The Compute Library is a set of computer vision and machine learning functions optimised for both Arm CPUs and GPUs using SIMD technologies. 

Adreno is not an Arm GPU, you will see little or no benefit from useing ACL.

Hope this helps.

**Comment by Saikatsaha1996** [comment]
> Hi @Saikatsaha1996
> 
> Quote from ACL documentation
> 
> > The Compute Library is a set of computer vision and machine learning functions optimised for both Arm CPUs and GPUs using SIMD technologies.
> 
> Adreno is not an Arm GPU, you will see little or no benefit from useing ACL.
> 
> Hope this helps.

Already run with termux in android..

And build also with same device..

![Screenshot_2022-04-02-20-17-56-085_com termux](https://user-images.githubusercontent.com/72664192/161500260-8602b63c-4961-4c0f-b346-2391062c4a7e.jpg)
![Screenshot_2022-04-02-20-10-19-505_com termux](https://user-images.githubusercontent.com/72664192/161500267-768d722a-a1fe-42a4-a74a-6f45969b35a0.jpg)
